### PR TITLE
Feature/req 56 module templates

### PR DESCRIPTION
**Pull request by viktorianer**
Modeling templates in the BSI IT baseline security allow to create and maintain the modules, safeguards and objects at freely definable, ideally at central location, and to use them in other points of application.

Using modelling templates, you can define such objects as modules or safeguards that provide reusable content. The content can be maintained at central or local location. It is not possible to change centrally maintained content at another point of application as central. Thereby, the responsibilities will be very finely regulated.

Thus information, descriptions and status can be maintained at one well-defined location and they will be displayed in real time at the required points.

### CODE REVIEW

File: sernet.gs.service/src/sernet/verinice/service/commands/LoadTemplatesOrImplementations.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+import org.apache.log4j.Logger;
+
+import sernet.gs.service.RetrieveInfo;
+import sernet.verinice.interfaces.CommandException;
+import sernet.verinice.interfaces.GenericCommand;
+import sernet.verinice.interfaces.IBaseDao;
+import sernet.verinice.model.common.CnATreeElement;
+
+/**
```
Comment (murygin) [inline sernet.gs.service/src/sernet/verinice/service/commands/LoadTemplatesOrImplementations.java]: Please add a simple and concise comment for this command.

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateActionDelegate.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import org.eclipse.ui.IObjectActionDelegate;
+
+import sernet.hui.common.VeriniceContext;
+import sernet.springclient.RightsServiceClient;
+import sernet.verinice.interfaces.ActionRightIDs;
+import sernet.verinice.interfaces.RightEnabledUserInteraction;
+import sernet.verinice.model.bsi.ITVerbund;
+import sernet.verinice.model.common.CnATreeElement;
+
+/**
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateActionDelegate.java]: Please add a simple and concise comment for this class.

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateBSIStrukturElementActionDelegate.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.ui.IWorkbenchPart;
+
+import sernet.gs.ui.rcp.main.ExceptionUtil;
+import sernet.gs.ui.rcp.main.common.model.CnAElementFactory;
+import sernet.gs.ui.rcp.main.common.model.CnAElementHome;
+import sernet.verinice.model.bsi.IBSIStrukturElement;
+import sernet.verinice.model.common.CnATreeElement;
+
+/**
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateBSIStrukturElementActionDelegate.java]: Please add a simple and concise comment for this class.

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateModuleActionDelegate.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.ui.IWorkbenchPart;
+
+import sernet.gs.ui.rcp.main.ExceptionUtil;
+import sernet.gs.ui.rcp.main.common.model.CnAElementFactory;
+import sernet.gs.ui.rcp.main.common.model.CnAElementHome;
+import sernet.verinice.model.bsi.BausteinUmsetzung;
+import sernet.verinice.model.common.CnATreeElement;
+
+/**
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateModuleActionDelegate.java]: Please add a simple and concise comment for this class.

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateSafeguardActionDelegate.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.ui.IWorkbenchPart;
+
+import sernet.gs.ui.rcp.main.ExceptionUtil;
+import sernet.gs.ui.rcp.main.common.model.CnAElementFactory;
+import sernet.gs.ui.rcp.main.common.model.CnAElementHome;
+import sernet.verinice.interfaces.CommandException;
+import sernet.verinice.model.bsi.MassnahmenUmsetzung;
+import sernet.verinice.model.common.CnATreeElement;
+
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateSafeguardActionDelegate.java]: Please add a simple and concise comment for this class.

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/views/TemplateView.java [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+import sernet.verinice.model.bsi.ITVerbund;
+import sernet.verinice.model.bsi.MassnahmenUmsetzung;
+import sernet.verinice.model.common.CnATreeElement;
+import sernet.verinice.model.iso27k.ISO27KModel;
+import sernet.verinice.rcp.IProgressRunnable;
+import sernet.verinice.rcp.RightsEnabledView;
+import sernet.verinice.rcp.TemplateSelectionDialog;
+import sernet.verinice.service.commands.LoadTemplatesOrImplementations;
+
+/**
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/views/TemplateView.java]: Please add a simple and concise comment for this view and for important methods in this view.

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/views/TemplateView.java [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+ * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.    
+ * See the GNU Lesser General Public License for more details.  
+ *  
+ * You should have received a copy of the GNU Lesser General Public License  
+ * along with this program.   
+ * If not, see <http://www.gnu.org/licenses/>.  
+ *   
+ * Contributors:  
+ *     Viktor Schmidt <vschmidt[at]ckc[dot]de> - initial API and implementation  
+ ******************************************************************************/
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/views/TemplateView.java]: Please create new package _sernet.verinice.rcp.templates_. Move all new RCP classes to this package.

File: sernet.gs.ui.rcp.main/src/sernet/verinice/iso27k/rcp/CopyTemplateElements.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+import org.apache.log4j.Logger;
+import org.eclipse.core.runtime.IProgressMonitor;
+
+import sernet.gs.ui.rcp.main.service.ServiceFactory;
+import sernet.verinice.interfaces.CommandException;
+import sernet.verinice.model.common.CnATreeElement;
+import sernet.verinice.service.commands.SaveElement;
+
+/**
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/iso27k/rcp/CopyTemplateElements.java]: Please add a simple and concise comment for this class.

File: sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateDecorator.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+ * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   
+ * See the GNU Lesser General Public License for more details. 
+ * 
+ * You should have received a copy of the GNU Lesser General Public License 
+ * along with this program.  
+ * If not, see <http://www.gnu.org/licenses/>. 
+ *  
+ * Contributors: 
+ *     Viktor Schmidt <vschmidt[at]ckc[dot]de> - initial API and implementation 
+ ******************************************************************************/
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateDecorator.java]: Please create new package sernet.verinice.rcp.templates. Move all new RCP classes to this package.

File: sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateSelectionComponent.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import org.eclipse.swt.widgets.Composite;
+
+import sernet.gs.ui.rcp.main.bsi.dialogs.Messages;
+import sernet.gs.ui.rcp.main.common.model.PlaceHolder;
+import sernet.gs.ui.rcp.main.service.ServiceFactory;
+import sernet.verinice.interfaces.CommandException;
+import sernet.verinice.model.common.CnATreeElement;
+import sernet.verinice.service.commands.LoadTemplateCandidates;
+
+/**
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateSelectionComponent.java]: Please add a simple and concise comment for this class.

File: sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateSelectionDialog.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.swt.widgets.Shell;
+
+import sernet.gs.ui.rcp.main.bsi.dialogs.CnATreeElementSelectionDialog;
+import sernet.gs.ui.rcp.main.bsi.dialogs.Messages;
+import sernet.verinice.model.common.CnATreeElement;
+
+/**
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateSelectionDialog.java]: Please add a simple and concise comment for this class.

File: sernet.gs.server/src/sernet/gs/server/DeleteOrphanTemplateRelationsJob.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+import org.quartz.JobExecutionContext;
+import org.quartz.JobExecutionException;
+import org.quartz.StatefulJob;
+import org.springframework.orm.hibernate3.HibernateCallback;
+import org.springframework.scheduling.quartz.QuartzJobBean;
+
+import sernet.verinice.interfaces.IBaseDao;
+import sernet.verinice.model.common.CnATreeElement;
+
+/**
```
Comment (viktorianer) [inline sernet.gs.server/src/sernet/gs/server/DeleteOrphanTemplateRelationsJob.java]: fixed

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/views/TemplateView.java (hunk 7 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/views/TemplateView.java]: fixed

File: sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateDecorator.java (hunk 9 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateDecorator.java]: fixed

File: sernet.gs.server/src/sernet/gs/server/DeleteOrphanTemplateRelationsJob.java (hunk 12 above)
Comment (viktorianer) [inline sernet.gs.server/src/sernet/gs/server/DeleteOrphanTemplateRelationsJob.java]: done

File: sernet.gs.server/src/sernet/gs/server/spring/veriniceserver-plain.xml [hunk 13]
```diff
@@ -62,6 +62,15 @@
   </property>
 </bean>
```
Comment (viktorianer) [inline sernet.gs.server/src/sernet/gs/server/spring/veriniceserver-plain.xml]: done

File: sernet.gs.service/src/sernet/verinice/model/common/CnATreeElement.java [hunk 14]
```diff
@@ -132,6 +131,11 @@ public void setNumericProperty(String propTypeId, int value) {
 	
 	private Set<Attachment> files = new HashSet<Attachment>(1);
```
Comment (viktorianer) [inline sernet.gs.service/src/sernet/verinice/model/common/CnATreeElement.java]: done

File: sernet.gs.service/src/sernet/verinice/service/commands/LoadOwnerByCnAElementDbId.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+import sernet.verinice.interfaces.GenericCommand;
+import sernet.verinice.interfaces.IBaseDao;
+import sernet.verinice.interfaces.bpm.IControlExecutionProcess;
+import sernet.verinice.model.common.ChangeLogEntry;
+
+/**
```
Comment (viktorianer) [inline sernet.gs.service/src/sernet/verinice/service/commands/LoadOwnerByCnAElementDbId.java]: deleted no longer needed class

File: sernet.gs.service/src/sernet/verinice/service/commands/LoadTemplateCandidates.java [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+import org.hibernate.criterion.Restrictions;
+
+import sernet.gs.service.RetrieveInfo;
+import sernet.verinice.interfaces.GenericCommand;
+import sernet.verinice.interfaces.IBaseDao;
+import sernet.verinice.model.bsi.BausteinUmsetzung;
+import sernet.verinice.model.bsi.MassnahmenUmsetzung;
+import sernet.verinice.model.common.CnATreeElement;
+
+/**
```
Comment (viktorianer) [inline sernet.gs.service/src/sernet/verinice/service/commands/LoadTemplateCandidates.java]: done

File: sernet.gs.service/src/sernet/verinice/service/commands/LoadTemplatesOrImplementations.java (hunk 1 above)
Comment (viktorianer) [inline sernet.gs.service/src/sernet/verinice/service/commands/LoadTemplatesOrImplementations.java]: done

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateActionDelegate.java (hunk 2 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateActionDelegate.java]: done

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateBSIStrukturElementActionDelegate.java (hunk 3 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateBSIStrukturElementActionDelegate.java]: done

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateModuleActionDelegate.java (hunk 4 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateModuleActionDelegate.java]: done

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateSafeguardActionDelegate.java (hunk 5 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/actions/MarkTemplateSafeguardActionDelegate.java]: done

File: sernet.gs.ui.rcp.main/src/sernet/verinice/iso27k/rcp/CopyTemplateElements.java (hunk 8 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/iso27k/rcp/CopyTemplateElements.java]: done

File: sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateSelectionComponent.java (hunk 10 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateSelectionComponent.java]: done

File: sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateSelectionDialog.java (hunk 11 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/TemplateSelectionDialog.java]: done

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/views/TemplateView.java (hunk 6 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/main/bsi/views/TemplateView.java]: done

File: sernet.gs.service/src/sernet/verinice/service/commands/CopyCommand.java [hunk 17]
```diff
@@ -429,6 +450,73 @@ private String getCopyTitle(final String title, final int n) {
         return Messages.getString("CopyCommand.0", title, n); //$NON-NLS-1$
     }
```
Comment (viktorianer) [inline sernet.gs.service/src/sernet/verinice/service/commands/CopyCommand.java]: done

File: sernet.gs.service/src/sernet/verinice/service/commands/CopyCommand.java [hunk 18]
```diff
@@ -281,10 +283,18 @@ private CnATreeElement saveCopy(CnATreeElement toGroup, CnATreeElement copyEleme
         copyElement = getDao().initializeAndUnproxy(copyElement);
         CnATreeElement newElement = saveNew(toGroup, copyElement);
         if(newElement.getEntity()!=null) {
-            newElement.getEntity().copyEntity(copyElement.getEntity());
```
Comment (murygin) [inline sernet.gs.service/src/sernet/verinice/service/commands/CopyCommand.java]: You can not copy a module template from am IT network to another IT network.

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/hibernate/CnATreeElement.hbm.xml [hunk 19]
```diff
@@ -121,6 +121,15 @@
         </set>
         
         <property name="iconPath" column="icon" type="string" length="256"/>
+        
+        <property name="templateTypeValue" type="string">
```
Comment (murygin) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/hibernate/CnATreeElement.hbm.xml]: After migrating an old DB to the new version I get this exception:

`java.lang.IllegalArgumentException: No enum constant sernet.verinice.model.common.CnATreeElement.TemplateType.none
	at java.lang.Enum.valueOf(Enum.java:238)
	at sernet.verinice.model.common.CnATreeElement$TemplateType.valueOf(CnATreeElement.java:1)
	at sernet.verinice.model.common.CnATreeElement.getTemplateType(CnATreeElement.java:877)
	at sernet.verinice.model.common.CnATreeElement.isImplementation(CnATreeElement.java:930)
	at sernet.gs.ui.rcp.main.bsi.views.TreeElementDecoratingLabelProvider.getForegroundColorForElement(TreeElementDecoratingLabelProvider.java:73)`

Change the default value to `default="'NONE'"`

**Comment by murygin** [comment]
Deleting a fresh import of a RECPLAST DB fails:
`ObjectDeletedException, deleted object would be re-saved by cascade (remove deleted object from associations): [sernet.verinice.model.bsi.risikoanalyse.GefaehrdungsUmsetzung#394270]
ObjectDeletedException, deleted object would be re-saved by cascade (remove deleted object from associations): [sernet.verinice.model.bsi.risikoanalyse.GefaehrdungsUmsetzung#394270]
  org.hibernate.impl.SessionImpl.forceFlush(SessionImpl.java:1014)
org.hibernate.event.def.DefaultSaveOrUpdateEventListener.entityIsTransient(DefaultSaveOrUpdateEventListener.java:165)
org.hibernate.event.def.DefaultSaveOrUpdateEventListener.performSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:94)
org.hibernate.event.def.DefaultSaveOrUpdateEventListener.onSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:70)
org.hibernate.impl.SessionImpl.fireSaveOrUpdate(SessionImpl.java:507)
org.hibernate.impl.SessionImpl.saveOrUpdate(SessionImpl.java:499)
org.hibernate.engine.CascadingAction$5.cascade(CascadingAction.java:218)
org.hibernate.engine.Cascade.cascadeToOne(Cascade.java:268)
org.hibernate.engine.Cascade.cascadeAssociation(Cascade.java:216)
org.hibernate.engine.Cascade.cascadeProperty(Cascade.java:169)
org.hibernate.engine.Cascade.cascadeCollectionElements(Cascade.java:296)
org.hibernate.engine.Cascade.cascadeCollection(Cascade.java:242)
org.hibernate.engine.Cascade.cascadeAssociation(Cascade.java:219)
org.hibernate.engine.Cascade.cascadeProperty(Cascade.java:169)
org.hibernate.engine.Cascade.cascade(Cascade.java:130)
org.hibernate.event.def.AbstractFlushingEventListener.cascadeOnFlush(AbstractFlushingEventListener.java:131)
org.hibernate.event.def.AbstractFlushingEventListener.prepareEntityFlushes(AbstractFlushingEventListener.java:122)
org.hibernate.event.def.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:65)
org.hibernate.event.def.DefaultAutoFlushEventListener.onAutoFlush(DefaultAutoFlushEventListener.java:35)
org.hibernate.impl.SessionImpl.autoFlushIfRequired(SessionImpl.java:969)
org.hibernate.impl.SessionImpl.list(SessionImpl.java:1562)
org.hibernate.impl.CriteriaImpl.list(CriteriaImpl.java:283)
org.springframework.orm.hibernate3.HibernateTemplate$36.doInHibernate(HibernateTemplate.java:1065)
org.springframework.orm.hibernate3.HibernateTemplate.doExecute(HibernateTemplate.java:419)
org.springframework.orm.hibernate3.HibernateTemplate.executeWithNativeSession(HibernateTemplate.java:374)
org.springframework.orm.hibernate3.HibernateTemplate.findByCriteria(HibernateTemplate.java:1055)
org.springframework.orm.hibernate3.HibernateTemplate.findByCriteria(HibernateTemplate.java:1048)
sernet.verinice.hibernate.HibernateDao.findByCriteria(HibernateDao.java:80)
sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:182)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:149)
org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:89)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:171)
org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
com.sun.proxy.$Proxy22.findByCriteria(Unknown Source)
sernet.verinice.service.commands.RemoveElement.isOrphanEntity(RemoveElement.java:317)
sernet.verinice.service.commands.RemoveElement.execute(RemoveElement.java:167)
sernet.verinice.service.HibernateCommandService.executeCommand(HibernateCommandService.java:151)
sun.reflect.GeneratedMethodAccessor281.invoke(Unknown Source)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:182)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:149)
org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:106)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:171)
org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
com.sun.proxy.$Proxy46.executeCommand(Unknown Source)
sun.reflect.GeneratedMethodAccessor281.invoke(Unknown Source)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:182)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:149)
org.springframework.remoting.support.RemoteInvocationTraceInterceptor.invoke(RemoteInvocationTraceInterceptor.java:77)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:171)
org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
com.sun.proxy.$Proxy48.executeCommand(Unknown Source)
sun.reflect.GeneratedMethodAccessor291.invoke(Unknown Source)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.springframework.remoting.support.RemoteInvocation.invoke(RemoteInvocation.java:205)
org.springframework.remoting.support.DefaultRemoteInvocationExecutor.invoke(DefaultRemoteInvocationExecutor.java:38)
org.springframework.remoting.support.RemoteInvocationBasedExporter.invoke(RemoteInvocationBasedExporter.java:78)
org.springframework.remoting.support.RemoteInvocationBasedExporter.invokeAndCreateResult(RemoteInvocationBasedExporter.java:114)
org.springframework.remoting.httpinvoker.HttpInvokerServiceExporter.handleRequest(HttpInvokerServiceExporter.java:74)
org.springframework.web.servlet.mvc.HttpRequestHandlerAdapter.handle(HttpRequestHandlerAdapter.java:49)
org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:875)
org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:807)
org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:571)
org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:511)
javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:487)
org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:362)
org.ops4j.pax.web.service.internal.HttpServiceServletHandler.handle(HttpServiceServletHandler.java:64)
org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:181)
org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:726)
org.ops4j.pax.web.service.internal.HttpServiceContext.handle(HttpServiceContext.java:111)
org.ops4j.pax.web.service.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:64)
org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
org.mortbay.jetty.Server.handle(Server.java:324)
org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:505)
org.mortbay.jetty.HttpConnection$RequestHandler.content(HttpConnection.java:842)
org.mortbay.jetty.HttpParser.parseNext(HttpParser.java:648)
org.mortbay.jetty.HttpParser.parseAvailable(HttpParser.java:211)
org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:380)
org.mortbay.io.nio.SelectChannelEndPoint.run(SelectChannelEndPoint.java:395)
org.mortbay.thread.BoundedThreadPool$PoolThread.run(BoundedThreadPool.java:450)
org.springframework.remoting.support.RemoteInvocationUtils.fillInClientStackTraceIfPossible(RemoteInvocationUtils.java:47)
org.springframework.remoting.support.RemoteInvocationResult.recreate(RemoteInvocationResult.java:115)
org.springframework.remoting.support.RemoteInvocationBasedAccessor.recreateRemoteInvocationResult(RemoteInvocationBasedAccessor.java:85)
org.springframework.remoting.httpinvoker.HttpInvokerClientInterceptor.invoke(HttpInvokerClientInterceptor.java:147)
org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:171)
org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
com.sun.proxy.$Proxy4.executeCommand(Unknown Source)
sernet.gs.ui.rcp.main.common.model.CnAElementHome.remove(CnAElementHome.java:260)
sernet.gs.ui.rcp.main.bsi.actions.DeleteHandler.removeElement(DeleteHandler.java:330)
sernet.gs.ui.rcp.main.bsi.actions.DeleteHandler$1.run(DeleteHandler.java:219)
org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:121)

  ObjectDeletedException, deleted object would be re-saved by cascade (remove deleted object from associations): [sernet.verinice.model.bsi.risikoanalyse.GefaehrdungsUmsetzung#394270]
`

File: sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/templates/TemplateView.java [hunk 20]
```diff
@@ -0,0 +1,518 @@
...
+import sernet.verinice.rcp.templates.TemplateTableViewer.PathCellLabelProvider;
+import sernet.verinice.service.commands.LoadTemplatesOrImplementations;
+
+/**
+ * This view shows for given ({@link CnATreeElement}) all modeling templates
+ * ({@link TemplateType#TEMPLATE}) applied (implemented) in this element, if the
+ * element is an implementation or all implementations
+ * ({@link TemplateType#IMPLEMENTATION}) that belong to this modeling template,
+ * if the element is a modeling template.
+ *
```
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/templates/TemplateView.java]: done

File: sernet.gs.server/src/sernet/gs/server/spring/veriniceserver-plain.xml [hunk 21]
```diff
@@ -100,6 +112,13 @@
     <property name="cronExpression" value="0 2 0 1/1 * ? *" />
 </bean>
 
+ <!-- Spring/Quartz trigger bean to execute cron job DeleteOrphanTemplateRelationsJob-->
+<bean id="deleteOrphanTemplateRelationsTrigger" class="org.springframework.scheduling.quartz.CronTriggerBean">
+	<property name="jobDetail" ref="deleteOrphanTemplateRelationsJob" />
```
Comment (viktorianer) [inline sernet.gs.server/src/sernet/gs/server/spring/veriniceserver-plain.xml]: fixed

File: sernet.gs.service/src/sernet/verinice/service/commands/RemoveElement.java [hunk 22]
```diff
@@ -159,6 +163,15 @@ public void execute() {
                 if (children[i] instanceof FinishedRiskAnalysis) {
                     removeRiskAnalysis((FinishedRiskAnalysis) children[i]);
                 }
+
```
Comment (viktorianer) [inline sernet.gs.service/src/sernet/verinice/service/commands/RemoveElement.java]: won't fix

File: sernet.gs.service/src/sernet/verinice/service/commands/CopyCommand.java (hunk 18 above)
Comment (viktorianer) [inline sernet.gs.service/src/sernet/verinice/service/commands/CopyCommand.java]: it is a feature of verinice core :)

File: sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/hibernate/CnATreeElement.hbm.xml (hunk 19 above)
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/gs/ui/rcp/hibernate/CnATreeElement.hbm.xml]: fixed

File: sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/templates/TemplateSelectionDialog.java [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+        final int cursorLocationYSubtrahend = 600;
+        newShell.setText(Messages.CnATreeElementSelectionDialog_2);
+        newShell.setSize(shellWidth, shellHeight);
+
+        // open the window right under the mouse pointer:
+        Point cursorLocation = Display.getCurrent().getCursorLocation();
+        newShell.setLocation(new Point(cursorLocation.x - cursorLocationXSubtrahend, cursorLocation.y - cursorLocationYSubtrahend));
+    }
+
+    @Override
```
Comment (viktorianer) [inline sernet.gs.ui.rcp.main/src/sernet/verinice/rcp/templates/TemplateSelectionDialog.java]: Remove unnecessary code duplicates

**Pull request merged**